### ISSUE TITLE
cloud_sql_proxy returns errors parsing config when running behind SOCKS5 proxy?

### ISSUE THREAD

**Issue by jackjyq** [GoogleCloudPlatform/cloudsql-proxy#1023]
## Question
cloud_sql_proxy returns `errors parsing config` when running behind SOCKS5 proxy.

## Additional Context

I am running  cloud_sql_proxy behind [ShadowsocksX-NG](https://github.com/shadowsocks/ShadowsocksX-NG/releases) at MacOS 12.0.1. I have set up a local socks5 proxy using shadowsocksX-NG.
<img width="592" alt="Screen Shot 2021-11-23 at 16 04 35" src="https://user-images.githubusercontent.com/16525413/142988825-0ab26f6d-4a1d-40c7-b214-d0da8a492219.png">

I have tested the socks5 proxy by curl, It seems OK.
```bash
➜  ~ curl -x socks5://127.0.0.1:1086 www.example.com
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
...
```
Then, I have configured the proxy using `gcloud config`

```ini
➜  ~ gcloud config list           
...
[proxy]
address = 127.0.0.1
port = 1086
type = socks5

Your active configuration is: [default]  
```
the gcloud info --run-diagnostics can pass, so does gcloud init

```bash
-> gcloud info --run-diagnostics
Network diagnostic detects and fixes local network connection issues.
Checking network connection...done.                                                                        
Reachability Check passed.
Network diagnostic passed (1/1 checks passed).

Property diagnostic detects issues that may be caused by properties.
Checking hidden properties...done.                                                                         
Hidden Property Check passed.
Property diagnostic passed (1/1 checks passed).
```

I follow the doc at [cloudsql-proxy](https://github.com/GoogleCloudPlatform/cloudsql-proxy#running-behind-a-socks5-proxy), it still not working.

```bash
-> ALL_PROXY=socks5://127.0.0.1:1086 cloud_sql_proxy -instances="xxx:us-west1:xxx"=tcp:3306
2021/11/23 16:12:05 Rlimits for file descriptors set to {&{8500 9223372036854775807}}
2021/11/23 16:12:37 errors parsing config:
	Get "https://sqladmin.googleapis.com/sql/v1beta4/projects/xxx/instances/xxx?alt=json&prettyPrint=false": dial tcp 172.217.163.42:443: i/o timeout 
```

```json
{
  "error": {
    "code": 401,
    "message": "Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.",
    "errors": [
      {
        "message": "Login Required.",
        "domain": "global",
        "reason": "required",
        "location": "Authorization",
        "locationType": "header"
      }
    ],
    "status": "UNAUTHENTICATED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "CREDENTIALS_MISSING",
        "domain": "googleapis.com",
        "metadata": {
          "method": "google.cloud.sql.v1beta4.SqlInstancesService.Get",
          "service": "sqladmin.googleapis.com"
        }
      }
    ]
  }
}
```

The `xxx` represents the instance/project name. 
My colleague has tried `cloud_sql_proxy -instances="xxx:us-west1:xxx"=tcp:3306` and it works fine.
Also, I have used `cloud_sql_proxy -instances="xxx:us-west1:xxx"=tcp:3306` before and it worked fine too.

Do you have an idea of the problem or how to gather more information to debug the issue?

**Comment by jackjyq** [comment]
I have read #361 and #769 and still can not figure out what's wrong? Please help.
I am wondering if there are some test cases for socks5 proxy?

**Comment by kurtisvg** [comment]
So this error:
```
2021/11/23 16:12:37 errors parsing config:
	Get "https://sqladmin.googleapis.com/sql/v1beta4/projects/xxx/instances/xxx?alt=json&prettyPrint=false": dial tcp 172.217.163.42:443: i/o timeout 
```
Means that it was unable to reach the SQL Admin API to determine the instance type. This is probably a problem with your SOCKS5 proxy? You could see if it also times out to manual curls to "https://sqladmin.googleapis.com/sql/v1beta4/"

It's not super clear where you got the second error from, but that error says that the API rejected your call because it was unauthenticated. This shouldn't happen with the proxy, as it will fail before it gets to that point. It could be that your SOCKS5 proxy isn't proxying the header correctly.

**Comment by jackjyq** [comment]
Hi @kurtisvg ,

Thanks for your reply, I have tried manual curls the ULR, it does NOT return time out.

```bash
➜  ~ curl -x socks5://127.0.0.1:1086 https://sqladmin.googleapis.com/sql/v1beta4/
<!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/sql/v1beta4/</code> was not found on this server.  <ins>That’s all we know.</ins>
```

Also, I have curled the following URLs as well, only one URL gives me an error:
- https://accounts.google.com
- https://cloudresourcemanager.googleapis.com/v1beta1/projects
- https://www.googleapis.com/auth/cloud-platform
- https://dl.google.com/dl/cloudsdk/channels/rapid/components-2.json

```bash
➜  curl -x socks5://127.0.0.1:1086 https://dl.google.com/dl/cloudsdk/channels/rapid/components-2.json
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to dl.google.com:443 
```
 
After digging into the LibreSSL SSL_connect: SSL_ERROR_SYSCALL issue. I tried `socks5h` and it returns me the json doc.

```bash
➜  curl -x socks5h://127.0.0.1:1086 https://dl.google.com/dl/cloudsdk/channels/rapid/components-2.json
```

Therefore, I use `socks5h` instead of `sock5` in order to use the remote DNS server.

```bash
➜  ALL_PROXY=socks5h://localhost:1086 cloud_sql_proxy -instances=xxx:us-west1:xxx=tcp:3306
2021/11/24 09:36:00 Rlimits for file descriptors set to {&{8500 9223372036854775807}}
2021/11/24 09:36:03 Listening on 127.0.0.1:3306 for xxx:us-west1:xxx
2021/11/24 09:36:03 Ready for new connections
```

It seems that the issue is caused by DNS, then I confirmed it with the following commands.

```bash
➜  ~ nslookup dl.google.com
Server:		114.114.114.114
Address:	114.114.114.114#53

Non-authoritative answer:
Name:	dl.google.com
Address: 203.208.50.33

➜  ~ nslookup dl.google.com 8.8.8.8
Server:		8.8.8.8
Address:	8.8.8.8#53

Non-authoritative answer:
Name:	dl.google.com
Address: 113.108.239.225
```
`114.114.114.114` is my local DNS server.

Since there is often a problem in resolving Google's hostname in China, and developers in China usually need the proxy to connect to Google Cloud, maybe it is better to use`socks5h` in the [example](https://github.com/GoogleCloudPlatform/cloudsql-proxy#running-behind-a-socks5-proxy)?

**Comment by kurtisvg** [comment]
My guess is that the SQLAdmin client we are using isn't repeating `ALL_PROXY`, but does respect `HTTPS_PROXY`. I'm going to tag in @enocom to look a little further and see what he finds.

**Comment by enocom** [comment]
Here's what I found:

1. All HTTP traffic is configured with [`http.FromEnvironment`](https://pkg.go.dev/net/http@go1.17.3#ProxyFromEnvironment) which uses `HTTP_PROXY` and `HTTPS_PROXY`. This is what the SQL Admin client uses.
2. All TCP traffic to and from a Cloud SQL instance uses [`proxy.FromEnvironment`](https://pkg.go.dev/golang.org/x/net/proxy#FromEnvironment) which uses `ALL_PROXY`. This is what the Proxy's [default dialer](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/main/proxy/proxy/client.go#L517) uses.

So in short, the SQL Admin API client requires an `HTTPS_PROXY` or `HTTP_PROXY` setting and the dialer itself requires an `ALL_PROXY` setting.

I'll update the README. In most case, people will probably want to use the same proxy for both `ALL_PROXY` and `HTTPS_PROXY`. Note that `HTTPS_PROXY` and `HTTP_PROXY` both accept `socks5` as a protocol.

**Comment by jackjyq** [comment]
Hi @enocom ,

Thanks for your explanation. I have read #1025. 

My local DNS server can not resolve Google's IP address correctly possibly due to the Great Firewall of China. Therefore, I am wondering if we can write `sock5h` somewhere in README.md which could save people time in debugging the DNS resolution issue?

**Comment by enocom** [comment]
There's an issue to add support for `socks5h` in Go's net/http package here: https://github.com/golang/go/issues/24135. A [related CL](https://go-review.googlesource.com/c/net/+/156518) suggests that `socks5` already does DNS resolution on the server, but I haven't tested this myself.

Based on my reading, the only allowed protocols at the moment are `socks5`, `https` and `http`.

**Comment by kurtisvg** [comment]
@jackjyq mentioned in their issue that they used it, so I think we're okay to mention it in the README.

**Comment by enocom** [comment]
Oh, I see. Yes, then we'll add it to the README. I'll update the PR.

**Comment by enocom** [comment]
Correction: `socks5h` is supported when using it with `ALL_PROXY`. My comment about was in reference to `HTTPS_PROXY`. 

https://cs.opensource.google/go/x/net/+/d83791d6:proxy/proxy.go;l=92;drc=d418f374d30933c6c7db22cf349625c295a5afaa

**Comment by kurtisvg** [comment]
@jackjyq can you confirm if you needed `socks5h`, or was just `HTTPS_PROXY` set enough?
